### PR TITLE
[Data Factory] Add data flow and debug session cmdlets

### PR DESCRIPTION
**Pull request by jackmagic313**
## Description

Design review tickets:
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/388
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/390

Changes:

* Adding CRUD commands for ADF V2 data flow: Set-AzDataFactoryV2DataFlow, Remove-AzDataFactoryV2DataFlow, and Get-AzDataFactoryV2DataFlow.
* Adding action commands for ADF V2 data flow debug Session: Start-AzDataFactoryV2DataFlowDebugSession, Get-AzDataFactoryV2DataFlowDebugSession, Add-AzDataFactoryV2DataFlowDebugSessionPackage, Invoke-AzDataFactoryV2DataFlowDebugSessionCommand and Stop-AzDataFactoryV2DataFlowDebugSession.
* Update ADF .Net SDK version to 4.2.0

## Checklist

- [x] I have read the [_Submitting Changes_](../CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/DataFactory/DataFactoryV2/DataFlowDebugSessions/InvokeAzureDataFactoryDataFlowDebugSessionCommandCommand.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                CommandPayload = new DataFlowDebugCommandPayload
+                {
+                    StreamName = StreamName,
+                    RowLimits = RowLimits,
+                    Expression = Expression,
+                    Columns = Columns
+                }
+
+            };
+            WriteObject(DataFactoryClient.InvokeDataFlowDebugSessionCommand(ResourceGroupName, DataFactoryName, request));
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/DataFlowDebugSessions/InvokeAzureDataFactoryDataFlowDebugSessionCommandCommand.cs]: Needs to actually implement ShouldProcess

File: src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StartAzureDataFactoryDataFlowDebugSessionCommand.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                TimeToLive = 60
+            };
+
+            if (!string.IsNullOrWhiteSpace(IntegrationRuntimeFile))
+            {
+                IntegrationRuntimeDebugResource integrationRuntimeResource = this.ConvertRequestFromJson(IntegrationRuntimeFile);
+                request.IntegrationRuntime = integrationRuntimeResource;
+            }
+
+            WriteObject(DataFactoryClient.StartDebugSession(ResourceGroupName, DataFactoryName, request));
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StartAzureDataFactoryDataFlowDebugSessionCommand.cs]: ShouldProcess implementation required

File: src/DataFactory/DataFactoryV2/DataFlows/RemoveAzureDataFactoryDataFlowCommand.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+using System.Globalization;
+using System.Management.Automation;
+using System.Net;
+using Microsoft.Azure.Commands.DataFactoryV2.Models;
+using Microsoft.Azure.Commands.DataFactoryV2.Properties;
+
+namespace Microsoft.Azure.Commands.DataFactoryV2
+{
+    [Cmdlet("Remove", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "DataFactoryV2DataFlow", DefaultParameterSetName = ParameterSetNames.ByFactoryName,SupportsShouldProcess = true), OutputType(typeof(void))]
+    public class RemoveAzureDataFactoryDataFlowCommand : DataFactoryContextActionBaseCmdlet
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/DataFlows/RemoveAzureDataFactoryDataFlowCommand.cs]: SHould implement a PassThru parameter that returns true here - then the return type will be bool

File: src/DataFactory/DataFactoryV2/DataFlows/SetAzureDataFactoryDataFlowCommand.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            {
+                ResourceGroupName = ResourceGroupName,
+                DataFactoryName = DataFactoryName,
+                Name = Name,
+                RawJsonContent = rawJsonContent,
+                Force = Force.IsPresent,
+                ConfirmAction = ConfirmAction
+            };
+
+            WriteObject(DataFactoryClient.CreatePSDataFlow(parameters));
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/DataFlows/SetAzureDataFactoryDataFlowCommand.cs]: Needs to implement ShouldProcess

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: You will need to provide `online version` for all of the new markdown files.  Follow the pattern for the existing files.

File: src/DataFactory/DataFactoryV2/help/Add-AzDataFactoryV2DataFlowDebugSessionPackage.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Add-AzDataFactoryV2DataFlowDebugSessionPackage.md]: You will need to provide `online version` for all of the new markdown files. Follow the pattern for the existing files.

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md]: same comment

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+## DESCRIPTION
+List all active data flow debug sessions by Azure Data Factory with details.
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\WINDOWS\system32> Get-AzDataFactoryV2DataFlowDebugSession -ResourceGroupName adf -DataFactoryName WikiADF
+
+DataFlowName           : DebugSession-0a7e0d6e-f2b7-48cc-8cd8-618326f5662f
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md]: Table format?

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+The Get-AzDataFactoryV2DataFlow cmdlet gets information about data flows in Azure Data Factory.
+If you specify the name of a data flow, this cmdlet gets information about that data flow.
+If you do not specify a name, this cmdlet gets information about all the data flows in the data factory.
+
+## EXAMPLES
+### Example 1: Get information about all data flows
+```powershell
+PS C:\> Get-AzDataFactoryV2DataFlow -ResourceGroupName "ADF" -DataFactoryName "WikiADF"
+
+DataFlowName      : TaxiDemo1
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: Table format?

File: src/DataFactory/DataFactoryV2/help/Add-AzDataFactoryV2DataFlowDebugSessionPackage.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+1. Start-AzDataFactoryV2DataFlowDebugSession
+1. Add-AzDataFactoryV2DataFlowDebugSessionPackage
+1. Invoke-AzDataFactoryV2DataFlowDebugSessionCommand (repeat this step for different commands/targets, or repeat step 2-3 in order to change the package file)
+1. Stop-AzDataFactoryV2DataFlowDebugSession
+
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\WINDOWS\system32> Add-AzDataFactoryV2DataFlowDebugSessionPackage -ResourceGroupName adf -DataFactoryName WikiADF -PackageFile "D:\dataflowps\addpackage.json" -SessionId 550effe4-93a3-485c-8525-eaf25259efbd
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Add-AzDataFactoryV2DataFlowDebugSessionPackage.md]: Should show the format

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md (hunk 5 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: See: https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/generate-format.ps1xml-file.md

File: src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md]: same comment

File: src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+## EXAMPLES
+
+### Example 1
+```powershell
+PS C:\WINDOWS\system32> $result = Invoke-AzDataFactoryV2DataFlowDebugSessionCommand -ResourceGroupName adf -DataFactoryName WiKiADF -Command executePreviewQuery -SessionId fd76cd0d-8b37-4dc0-a370-3f9d43ac686d -StreamName source1 -RowLimits 100 -AsJob
+PS C:\WINDOWS\system32> $result
+
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+3      Long Running... AzureLongRun... Running       True            localhost            Invoke-AzDataFactoryV2...
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md]: nice

File: src/DataFactory/DataFactoryV2/help/Remove-AzDataFactoryV2DataFlow.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Remove-AzDataFactoryV2DataFlow.md]: same comment

File: src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md [hunk 14]
```diff
@@ -0,0 +1,202 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md]: same comment

File: src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+## DESCRIPTION
+The Set-AzDataFactoryV2DataFlow cmdlet creates a data flow or updates an existing data flow in Azure Data Factory.
+
+## EXAMPLES
+
+### Example 1: Create a data flow
+```powershell
+PS C:\> Set-AzDataFactoryV2DataFlow -ResourceGroupName "ADF" -DataFactoryName "WikiADF" -Name "TaxiDemo1" -DefinitionFile "C:\\samples\\WikiSample\\TaxiDemo1.json"
+
+DataFlowName      : TaxiDemo1
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md]: table format?

File: src/DataFactory/DataFactoryV2/help/Start-AzDataFactoryV2DataFlowDebugSession.md [hunk 16]
```diff
@@ -0,0 +1,235 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Start-AzDataFactoryV2DataFlowDebugSession.md]: need online version

File: src/DataFactory/DataFactoryV2/help/Stop-AzDataFactoryV2DataFlowDebugSession.md [hunk 17]
```diff
@@ -0,0 +1,211 @@
+---
+external help file: Microsoft.Azure.PowerShell.Cmdlets.DataFactoryV2.dll-Help.xml
+Module Name: Az.DataFactory
+online version:
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Stop-AzDataFactoryV2DataFlowDebugSession.md]: online version

**Comment by jackmagic313** [comment]
Thanks @markcowl for the quick review! I just addressed the feedbacks.

File: src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StopAzureDataFactoryDataFlowDebugSessionCommand.cs [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+using Microsoft.Azure.Commands.DataFactoryV2.Properties;
+using System.Globalization;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.DataFactoryV2
+{
+    [Cmdlet("Stop", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "DataFactoryV2DataFlowDebugSession", DefaultParameterSetName = ParameterSetNames.ByFactoryName, SupportsShouldProcess = true)]
+    [OutputType(typeof(void))]
+    [OutputType(typeof(string))]
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StopAzureDataFactoryDataFlowDebugSessionCommand.cs]: It outputs bool

File: src/DataFactory/DataFactoryV2/DataFlows/RemoveAzureDataFactoryDataFlowCommand.cs (hunk 3 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/DataFlows/RemoveAzureDataFactoryDataFlowCommand.cs]: Why is this returning string?

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md (hunk 9 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: This is list format

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+Properties        : Microsoft.Azure.Management.DataFactory.Models.MappingDataFlow
+```
+
+This command gets information about all data flows in the data factory named WikiADF.
+
+### Example 2: Get information about a specific data flow
+```powershell
+PS C:\> Get-AzDataFactoryV2DataFlow -ResourceGroupName "ADF" -DataFactoryName "WikiADF" -Name "dataflow1"
+
+DataFlowName      : dataflow1
```
Comment (markcowl) [inline src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: list format.  You will need to have a format entry for this in order to have a table format, see: https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/generate-format.ps1xml-file.md

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md (hunk 8 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlowDebugSession.md]: still lidt format

File: src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md (hunk 12 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/help/Invoke-AzDataFactoryV2DataFlowDebugSessionCommand.md]: (this is table format)

File: src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md (hunk 15 above)
Comment (markcowl) [reply src/DataFactory/DataFactoryV2/help/Set-AzDataFactoryV2DataFlow.md]: this is list format

File: src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md (hunk 19 above)
Comment (jackmagic313) [reply src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: Oh I see. I thought I need to include a example to show how to convert to a table format result 😅, will update my pr soon

Comment (jackmagic313) [reply src/DataFactory/DataFactoryV2/help/Get-AzDataFactoryV2DataFlow.md]: The generation tool doesn't work with "Could not load file or assembly 'System.Runtime.Loader, Version=4.0.0.0, Culture=neutral,
PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified." error (I tried to change the version to 4.0.0 in csproj and I got a different error with something about assembly context). For now I just added manually into format file instead and it should still work if someone uses the script to generate the file in the future.

File: src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StopAzureDataFactoryDataFlowDebugSessionCommand.cs (hunk 18 above)
Comment (jackmagic313) [reply src/DataFactory/DataFactoryV2/DataFlowDebugSessions/StopAzureDataFactoryDataFlowDebugSessionCommand.cs]: not sure what I did :P it's fixed.

**Pull request merged**